_posts/2019-11-04-generic-programming-part-2-implementation-overview.markdown

```diff
@@ -0,0 +1,603 @@
+---
+layout: post
+title:  "Generic Programming - Implementation Overview"
+date:   2019-11-04 12:00:00 +0300
+---
+
+![Types of Asteraceae](https://media.githubusercontent.com/media/VysotskiVadim/VysotskiVadim.github.io/a9ae80c482ebd479ccbc702e9a8ac3d67b369ecb/assets/Asteracea_poster_3_part_2.jpg){: style="width:100%"}
+
+During my career I've been working with different strongly typed languages that support generic programming.
+They faced the same set of challenges but solved it in different time using different methods with different pros and cons.
+I find it fascinating, that's why I blogged about it.
+
+If you don't have much experience with generic programming -
+read [the introduction to generic programming]({% post_url 2019-10-01-generic-programming-part-1-introduction %}).
+
+For every considered language we will answer following questions:
+* How does generics work under the hood?
+* How did migration to generic happen?
+* [Variance]({% post_url 2019-10-01-generic-programming-part-1-introduction %}#variance)
+* Pros and Cons
+
+I won't compare languages in order to understand which one is better,
+because they all are nice!
+My point is to show how different preconditions like language and runtime design,
+ecosystem, or even market competition affected the way how generics were implemented.
+
+## C++ {#cpp}
+
+In C++ developers control all aspects of memory usage:
+you can allocate objects on stack or heap,
+you're in charge of freeing memory,
+you can even develop your onw heap allocator.
+Decisions of how memory would be allocated are made in the place of usage:
+
+```cpp
+class Product {
+    private:
+        double price;
+    public:
+        Product(double price): price(price) { }
+        void printPrice() {
+            std::cout << "my price is " << price << "\n";
+        }
+    };
+ 
+   int main()
+   {
+        Product stackAllocatedProduct(1.5);
+        stackAllocatedProduct.printPrice();
+        Product* heapAllocatedProduct = new Product(2.5);
+        heapAllocatedProduct->printPrice();
+        return 0;
+    }
+```
+
+In C++ methods are invoked by 2 different ways.
+If method declared using `virtual` keyword it's invoked via [virtual method table](https://en.wikipedia.org/wiki/Virtual_method_table).
+Other methods are invoked directly: i.e. compiler puts in machine code call function by its address in memory.
```

Comment:
compiler puts in machine code call function by its address =>  compiler puts in machine code function call by its address?

Reply:
nice one!